File: backend/simulation.py
import random

class Cell:
    def __init__(self, color=[106, 103, 78]):
        self.state = "EMPTY"
        self.age = 0
        self.color = color

class Simulation:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        self.MAX_AGE = 150
        self.DECAY_DURATION = 25

        self.grid = [[Cell() for _ in range(self.width)] for _ in range(self.height)]

        for row in self.grid:
            for cell in row:
                if random.random() < 0.35:
                    cell.state = "ALIVE"

    def tick(self):
        new_grid = [[Cell() for _ in range(self.width)] for _ in range(self.height)]

        for row in range(self.height):
            for col in range(self.width):

                alive_neighbors = 0
                decaying_neighbors = 0

                for dr in range(-1, 2):
                    for dc in range(-1, 2):
                     if dr == 0 and dc == 0:
                         continue

                     nr = (row + dr) % self.height
                     nc = (col + dc) % self.width

                     neighbor_cell = self.grid[nr][nc]
                     if neighbor_cell.state in ["ALIVE", "REBORN"]:
                         alive_neighbors += 1
                         
                     elif neighbor_cell.state == "DECAYING":
                         decaying_neighbors += 1

                current_cell = self.grid[row][col]
                new_cell = new_grid[row][col]

                if current_cell.state in ["ALIVE", "REBORN"]:
                    if current_cell.age > self.MAX_AGE:
                        new_cell.state = "DECAYING"
                        new_cell.age = 0                        
                    elif alive_neighbors < 2 or alive_neighbors > 3:
                        new_cell.state = "EMPTY"
                    else:
                        new_cell.state = current_cell.state
                        new_cell.color = current_cell.color
                        new_cell.age = current_cell.age + 1
                        
                elif current_cell.state == "EMPTY":
                    if alive_neighbors == 3:
                        new_cell.state = "ALIVE"
                        for dr in range(-1, 2):
                            for dc in range(-1, 2):
                                if dr == 0 and dc == 0: continue

                                parent = self.grid[(row + dr) % self.height][(col + dc) % self.width]
                                if parent.state in ["ALIVE", "REBORN"]:
                                    new_cell.color = parent.color
                                    break
                                
                    elif decaying_neighbors > 0:
                        birth_chance = 0.15 * decaying_neighbors
                        if random.random() < birth_chance:
                            new_cell.state = "REBORN"
                            new_cell.color = [random.randint(50, 200) for _ in range(3)]
                                                
                elif current_cell.state == "DECAYING":
                    if current_cell.age > self.DECAY_DURATION:
                        new_cell.state = "EMPTY"
                    else:
                        new_cell.state = "DECAYING"
                        new_cell.age = current_cell.age + 1
                        
        self.grid = new_grid

    def get_cell_states(self):
        grid_data = []
        for r in range(self.height):
            row_data = []
            for c in range(self.width):
                cell = self.grid[r][c]
                row_data.append({
                                    "state": cell.state,
                                    "age": cell.age,
                                    "color": cell.color
                                })
            grid_data.append(row_data)
        return grid_data

File: frontend/src/simulation.js
export class Cell {
  constructor(state = "EMPTY", color = [106, 103, 78], age = 0) {
    this.state = state;
    this.color = color;
    this.age = age;
  }
}

// --- Simulation parameters ---
export const WIDTH = 500;
export const HEIGHT = 500;
export const MAX_AGE = 150;
const DECAY_DURATION = 25;
const INITIAL_DENSITY = 0.35;

// --- Module-level state ---
let grid = [];
let activeCells = new Set();
let isInitialized = false;

// --- Private helper functions ---
function createEmptyGrid() {
  return Array.from({ length: HEIGHT }, () =>
    Array.from({ length: WIDTH }, () => new Cell())
  );
}

// Initialize the simulation with random grid
export function init() {
  grid = createEmptyGrid();
  activeCells.clear();
  for (let row = 0; row < HEIGHT; row++) {
    for (let col = 0; col < WIDTH; col++) {
      if (Math.random() < INITIAL_DENSITY) {
        grid[row][col].state = "ALIVE";
        activeCells.add(`${row},${col}`);
      }
    }
  }
  isInitialized = true;
  console.log("Client-side simulation initialized");
}

// Core simulation logic
export function tick() {
  if (!isInitialized || activeCells.size === 0) return;

  const cellsToEvaluate = new Set();
  for (const cell of activeCells) {
    const [row, col] = cell.split(',').map(Number);
    for (let drow = -1; drow <= 1; drow++) {
      for (let dcol = -1; dcol <= 1; dcol++) {
        const nrow = (row + drow + HEIGHT) % HEIGHT;
        const ncol = (col + dcol + WIDTH) % WIDTH;
        cellsToEvaluate.add(`${nrow},${ncol}`);
      }
    }
  }

  const changes = [];

  for (const cell of cellsToEvaluate) {
    const [row, col] = cell.split(',').map(Number);
    
    let aliveNeighbors = 0;
    let decayingNeighbors = 0;
    let parentColor = null;

    // Count neighbors
    for (let drow = -1; drow <= 1; drow++) {
      for (let dcol = -1; dcol <= 1; dcol++) {
        if (drow === 0 && dcol === 0) continue;

        const nrow = (row + drow + HEIGHT) % HEIGHT;
        const ncol = (col + dcol + WIDTH) % WIDTH;

        const neighbor = grid[nrow][ncol];
        if (neighbor.state === "ALIVE" || neighbor.state === "REBORN") {
          aliveNeighbors++;
          if (!parentColor) {
            parentColor = neighbor.color;
          }
        } else if (neighbor.state === "DECAYING") {
          decayingNeighbors++;
        }
      }
    }

    const currentCell = grid[row][col];
    const newCell = new Cell(currentCell.state, currentCell.color, currentCell.age);

    if (currentCell.state === "ALIVE" || currentCell.state === "REBORN") {
      if (currentCell.age > MAX_AGE) {
        newCell.state = "DECAYING";
        newCell.age = 0; // Reset age for decay
      } else if (aliveNeighbors < 2 || aliveNeighbors > 3) {
        newCell.state = "EMPTY";
        newCell.age = 0;
      } else {
        newCell.age++;
      }
    } else if (currentCell.state === "EMPTY") {
      if (aliveNeighbors === 3) {
        newCell.state = "ALIVE";
        newCell.color = parentColor || [106, 103, 78];
        newCell.age = 0;
      } else if (decayingNeighbors > 0) {
        const birthChance = 0.15 * decayingNeighbors;
        if (Math.random() < birthChance) {
          newCell.state = "REBORN";
          newCell.color = Array.from({ length: 3 }, () => Math.random() * 150 + 50);
          newCell.age = 0;
        }
      }
    } else if (currentCell.state === "DECAYING") {
      if (currentCell.age > DECAY_DURATION) {
        newCell.state = "EMPTY";
        newCell.age = 0;
      } else {
        newCell.age++;
      }
    }
    
    if (newCell.state !== currentCell.state || newCell.age !== currentCell.age) {
        changes.push({row, col, cell: newCell});
    }
  }

  const newActiveCells = new Set();
  for(const change of changes) {
      grid[change.row][change.col] = change.cell;
      if(change.cell.state !== 'EMPTY') {
          newActiveCells.add(`${change.row},${change.col}`);
      }
  }
  activeCells = newActiveCells;
}

export function shiftGrid(dx, dy) {
  if (!isInitialized || (dx === 0 && dy === 0)) return;

  const newActiveCells = new Set();
  for (const cell of activeCells) {
    const [row, col] = cell.split(',').map(Number);
    const newRow = (row - dy + HEIGHT) % HEIGHT;
    const newCol = (col - dx + WIDTH) % WIDTH;
    newActiveCells.add(`${newRow},${newCol}`);
  }
  activeCells = newActiveCells;

  if (dy < 0) {
    for (let i = 0; i < -dy; i++) {
      grid.unshift(grid.pop());
    }
  }
  else if (dy > 0) {
    for (let i = 0; i < dy; i++) {
      grid.push(grid.shift());
    }
  }

  if (dx !== 0) {
    for (const row of grid) {
      if (dx < 0) {
        for (let i = 0; i < -dx; i++) {
          row.unshift(row.pop());
        }
      }
      else if (dx > 0) {
        for (let i = 0; i < dx; i++) {
          row.push(row.shift());
        }
      }
    }
  }
}

export function getSimulationState() {
  return {
    grid,
    MAX_AGE,
  };
}
